README.md
# AmazonProductAPI
PHP library to perform product lookup and searches using the Amazon Product API.

## Installation

This library requires the [SimpleXML](http://php.net/manual/en/book.simplexml.php) and [Curl](http://php.net/manual/en/book.curl.php) extensions to be installed and uses PHP 7+ . Installation is simple using [Composer](https://composer.io):

```shell
composer require marcl/amazonproductapi
```

### Amazon Product API
It also assumes that you have some basic knowledge of Amazon's Product API and have set up an Amazon Associate account see: [Amazon Product API Set Up](http://docs.amazonwebservices.com/AWSECommerceService/2011-08-01/GSG/GettingSetUp.html).

You'll need an AWS key, secret key, and associate tag. Ensure that you keep these safe!

## Examples

I've added some simple examples in `examples.php`. To run them create a file called `secretKeys.php` containing your secret keys:

```php
<?php
$keyId = 'YOUR-AWS-KEY';
$secretKey = 'YOUR-AWS-SECRET-KEY';
$associateId = 'YOUR-AMAZON-ASSOCIATE-ID';
?>
```

and then run the examples with:

```shell
php examples.php
```

## Quick Start

Include the library in your code using the Composer autoloader and create an AmazonUrlBuilder with your credentials

```php
require('vendor/autoload.php');

use MarcL\AmazonAPI;
use MarcL\AmazonUrlBuilder;

// Keep these safe
$keyId = 'YOUR-AWS-KEY';
$secretKey = 'YOUR-AWS-SECRET-KEY';
$associateId = 'YOUR-AMAZON-ASSOCIATE-ID';

// Setup a new instance of the AmazonUrlBuilder with your keys
$urlBuilder = new AmazonUrlBuilder(
    $keyId,
    $secretKey,
    $associateId,
    'uk'
);

// Setup a new instance of the AmazonAPI and define the type of response
$amazonAPI = new AmazonAPI($urlBuilder, 'simple');

$items = $amazonAPI->ItemSearch('harry potter', 'Books', 'price');

```

**Note:** Keep your Amazon keys safe. Either use environment variables or include from a file that you don't check into GitHub.

### Locale

This library supports all [Product Advertising API locales](http://docs.aws.amazon.com/AWSECommerceService/latest/DG/Locales.html) and you can set it as you construct the AmazonUrlBuilder class with your keys.

At this time, these are the current supported locales:

* Brazil ('br')
* Canada ('ca')
* China ('cn')
* France ('fr')
* Germany ('de')
* India ('in')
* Italy ('it')
* Japan ('jp')
* Mexico ('mx')
* Spain ('es')
* United Kingdom ('uk')
* United States ('us')

### Item Search
To search for an item use the `ItemSearch()` method:

```php
// Search for harry potter items in all categories
$items = $amazonAPI->ItemSearch('harry potter');

// Search for harry potter items in Books category only
$items = $amazonAPI->ItemSearch('harry potter', 'Books');
```

#### Default sort

By default, the `ItemSearch()` method will search by featured. If you want to sort by another category then pass a 3rd parameter with the name of the category you wish to sort by. These differ by category type but the two you'll probably need are `price` (sort by price low to high) or `-price` (sort by price high to low). See [ItemSearch Sort Values](http://docs.aws.amazon.com/AWSECommerceService/latest/DG/APPNDX_SortValuesArticle.html) for more details.

```php
// Search for harry potter items in Books category, sort by low to high
$items = $amazonAPI->ItemSearch('harry potter', 'Books', 'price');

// Search for harry potter items in Books category, sort by high to low
$items = $amazonAPI->ItemSearch('harry potter', 'Books', '-price');
```

To determine valid categories for search call `GetValidSearchNames()`:

```php
// Get an array of valid search categories we can use
$searchCategories = $amazonAPI->GetValidSearchNames();
```

### Item Lookup
To look up product using the product ASIN number use `ItemLookup()`:

```php
// Retrieve specific item by id
$items = $amazonAPI->ItemLookUp('B003U6I396');

// Retrieve a list of items by ids
$asinIds = array('B003U6I396', 'B003U6I397', 'B003U6I398');
$items = $amazonAPI->ItemLookUp($asinIds);
```

### Data Transformation
By default the data will be returned as SimpleXML nodes. However, you can ask for the data to be transformed differently, depending on your use case for the API. Pass a type when instantiating the AmazonAPI class as follows:

```php
// Default return type is XML
$amazonAPI = new AmazonAPI($amazonUrlBuilder);
$items = $amazonAPI->ItemSearch('harry potter');
var_dump($items);
```

This will output:

```shell
class SimpleXMLElement#2 (2) {
	public $OperationRequest =>
		class SimpleXMLElement#3 (3) {
			public $RequestId =>
			string(36) "de58449e-0c1a-47ac-9823-00fd049c52df"
			public $Arguments =>
			class SimpleXMLElement#5 (1) {
				public $Argument =>
				array(11) {
	...
```

```php
// Return simplified data
$amazonAPI = new AmazonAPI($amazonUrlBuilder, 'simple');
$items = $amazonAPI->ItemSearch('harry potter');
var_dump($items);
```

This will return a simplified version of each item with minimal data but enough for simple use cases.

```
array(10) {
	[0] =>
	array(8) {
	'asin' =>
	string(10) "B00543R3WG"
	'url' =>
	string(212) "http://www.amazon.co.uk/Harry-Potter-Complete-8-Film-Collection/dp/B00543R3WG%3FSubscriptionId%3D1BM0B8TXM1YSZ1M0XDR2%26tag%3Ddjcr-21%26linkCode%3Dxm2%26camp%3D2025%26creative%3D165953%26creativeASIN%3DB00543R3WG"
	'rrp' =>
	double(44.99)
	'title' =>
	string(58) "Harry Potter - The Complete 8-Film Collection [DVD] [2011]"
	'lowestPrice' =>
	double(23.4)
	'largeImage' =>
	string(53) "http://ecx.images-amazon.com/images/I/51qa9nTUsEL.jpg"
	'mediumImage' =>
	string(61) "http://ecx.images-amazon.com/images/I/51qa9nTUsEL._SL160_.jpg"
	'smallImage' =>
	string(60) "http://ecx.images-amazon.com/images/I/51qa9nTUsEL._SL75_.jpg"
	}
	[1] =>
	array(8) {
	'asin' =>
	string(10) "0747558191"
	'url' =>
	string(212) "http://www.amazon.co.uk/Harry-Potter-Philosophers-Stone-Rowling/dp/0747558191%3FSubscriptionId%3D1BM0B8TXM1YSZ1M0XDR2%26tag%3Ddjcr-21%26linkCode%3Dxm2%26camp%3D2025%26creative%3D165953%26creativeASIN%3D0747558191"
	'rrp' =>
	double(6.99)
	'title' =>
	string(40) "Harry Potter and the Philosopher\'s Stone"
	…
```

The different data transformation types are defined as follows. Feel free to raise an issue if you'd like the data transforming to a new type.

* **xml** - (Default) returns data as SimpleXML nodes.
* **array** - Returns data as PHP arrays and objects.
* **simple** - Returns data as simplified arrays and doesn't contain all API data. Use this if you just need prices, title and images.
* **json** - Returns data as a JSON string. Use this for returning from a server API endpoint.

## TODO

* Need to make the simplified data less hardcoded!

## Thanks

This library uses code based on [AWS API authentication For PHP](http://randomdrake.com/2009/07/27/amazon-aws-api-rest-authentication-for-php-5/) by [David Drake](https://github.com/randomdrake) but has been mostly rewritten.

## LICENSE

See [LICENSE](LICENSE)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [3.0.1] - 2017-08-21
### Changed
- Fix bug where called data transformation wasn't used

## [3.0.0] - 2017-07-20
### Changed
- Total rewrite to make it composer friendly

## [2.0.0] - 2017-07-11
### Added
- Allow `ItemSearch` to be sorted by value

### Change
- Defaults to using HTTPS endpoint rather than HTTP

## [1.0.0] - 2017-07-11
### Added
- Initial version since 2012



src/AmazonUrlBuilder.php
<?php

namespace MarcL;

class AmazonUrlBuilder {
    private $secretKey = NULL;
    private $keyId = NULL;
    private $associateTag = NULL;
    private $amazonEndpoint = NULL;

	private $endpoint = 'onca/xml';

	private $localeTable = array(
		'br' => 'webservices.amazon.br/onca/xml',
		'ca' =>	'webservices.amazon.ca/onca/xml',
		'cn' =>	'webservices.amazon.cn/onca/xml',
		'fr' =>	'webservices.amazon.fr/onca/xml',
		'de' =>	'webservices.amazon.de/onca/xml',
		'in' =>	'webservices.amazon.in/onca/xml',
		'it' =>	'webservices.amazon.it/onca/xml',
		'jp' =>	'webservices.amazon.co.jp/onca/xml',
		'mx' =>	'webservices.amazon.com.mx/onca/xml',
		'es' =>	'webservices.amazon.es/onca/xml',
		'uk' =>	'webservices.amazon.co.uk/onca/xml',
		'us' =>	'webservices.amazon.com/onca/xml'
	);

	private function throwIfNull($parameterValue, $parameterName) {
		if ($parameterValue == NULL) {
			throw new \Exception($parameterName . ' should be defined');
		}
	}

    public function __construct($keyId, $secretKey, $associateTag, $locale = 'us') {
		$this->throwIfNull($keyId, 'Amazon key ID');
		$this->throwIfNull($secretKey, 'Amazon secret key');
		$this->throwIfNull($associateTag, 'Amazon associate tag');

        $this->secretKey = $secretKey;
        $this->amazonEndpoint = $this->GetAmazonApiEndpoint($locale);
        $this->associateTag = $associateTag;
        $this->keyId = $keyId;
    }

    private function GetAmazonApiEndpoint($locale) {
		if (!array_key_exists($locale, $this->localeTable)) {
			$locale = 'us';
		}

        return('https://' . $this->localeTable[$locale]);
    }

	private function CreateUnsignedAmazonUrl($params) {
		$baseParams = array(
			'Service' => 'AWSECommerceService',
			'AssociateTag' => $this->associateTag,
			'AWSAccessKeyId' => $this->keyId
		);

		$buildParams = array_merge($baseParams, $params);

		$request = $this->amazonEndpoint . '?' .http_build_query($buildParams);

		return($request);
	}

	private function createSignature($signatureString) {
	    return urlencode(
			base64_encode(
				hash_hmac(
					'sha256',
					$signatureString,
					$this->secretKey,
					true
				)
			)
		);
	}

	/**
	  * This function will take an existing Amazon request and change it so that it will be usable
	  * with the new authentication.
	  *
	  * @param string $request - your existing request URI
	  * @param string $version - (optional) the version of the service you are using
	  *
	  * @link http://www.ilovebonnie.net/2009/07/27/amazon-aws-api-rest-authentication-for-php-5/
	  */
	private function CreateSignedAwsRequest($request, $version = '2011-08-01') {
	    // Get a nice array of elements to work with
	    $uri_elements = parse_url($request);

	    // Grab our request elements
	    $request = $uri_elements['query'];

	    // Throw them into an array
	    parse_str($request, $parameters);

	    // Add the new required paramters
	    $parameters['Timestamp'] = gmdate("Y-m-d\TH:i:s\Z");
	    $parameters['Version'] = $version;

	    // The new authentication requirements need the keys to be sorted
	    ksort($parameters);

	    // Create our new request
	    foreach ($parameters as $parameter => $value) {
	        // We need to be sure we properly encode the value of our parameter
	        $parameter = str_replace("%7E", "~", rawurlencode($parameter));
	        $value = str_replace("%7E", "~", rawurlencode($value));
	        $requestArray[] = $parameter . '=' . $value;
	    }

	    // Put our & symbol at the beginning of each of our request variables and put it in a string
	    $requestParameters = implode('&', $requestArray);

	    // Create our signature string
	    $signatureString = "GET\n{$uri_elements['host']}\n{$uri_elements['path']}\n{$requestParameters}";
	    $signature = $this->createSignature($signatureString);

	    // Return our new request
		$newUrl = "http://{$uri_elements['host']}{$uri_elements['path']}?{$requestParameters}&Signature={$signature}";
	    return $newUrl;
	}

    public function generate($params) {
        $unsignedRequest = $this->CreateUnsignedAmazonUrl($params);

		return $this->CreateSignedAwsRequest($unsignedRequest);
    }
}

?>